src/mpl/include/mpl_gpu.h

```diff
@@ -0,0 +1,61 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/* vim: set ft=c.mpich : */
+/*
+ *  (C) 2020 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef MPL_GPU_H_INCLUDED
+#define MPL_GPU_H_INCLUDED
+
+#include "mplconfig.h"
+
+#ifdef MPL_HAVE_GPU
+
+#define MPL_GPU_SUCCESS 0
+#define MPL_GPU_FAILURE 1
+
+typedef enum {
+    MPL_GPU_POINTER_HOST = 0,
+    MPL_GPU_POINTER_DEV,
+    MPL_GPU_POINTER_MANAGED
+} MPL_pointer_type_t;
+
+int MPL_gpu_query_pointer_type(const void *ptr, MPL_pointer_type_t * attr);
+int MPL_gpu_memcpy_h2d(void *dst, const void *src, size_t count);
+int MPL_gpu_memcpy_d2h(void *dst, const void *src, size_t count);
+int MPL_gpu_memcpy_d2d(void *dst, const void *src, size_t count);
+
+int MPL_gpu_memcpy_h2d_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream);
+int MPL_gpu_memcpy_d2h_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream);
+int MPL_gpu_memcpy_d2d_async(void *dst, const void *src, size_t count, MPL_gpu_stream_t stream);
+
+int MPL_gpu_ipc_get_mem_handle(MPL_gpu_ipc_mem_handle_t * h_mem, void *ptr);
+int MPL_gpu_ipc_open_mem_handle(void *ptr, MPL_gpu_ipc_mem_handle_t * h_mem);
+int MPL_gpu_ipc_close_mem_handle(void *ptr);
+
+int MPL_gpu_ipc_get_event_handle(MPL_gpu_ipc_event_handle_t * h_event, MPL_gpu_event_t * event);
+int MPL_gpu_ipc_open_event_handle(MPL_gpu_event_t * event, MPL_gpu_ipc_event_handle_t * h_event);
+
+int MPL_gpu_event_create(MPL_gpu_event_t * event, int flag);
+int MPL_gpu_event_sync(MPL_gpu_event_t * event);
+int MPL_gpu_event_record(MPL_gpu_event_t * event);
+int MPL_gpu_event_destroy(MPL_gpu_event_t * event);
+
+int MPL_gpu_malloc(void **ptr, size_t size);
+int MPL_gpu_malloc_host(void **ptr, size_t size);
+int MPL_gpu_malloc_managed(void **ptr, size_t size, unsigned int flags);
+int MPL_gpu_free(void *ptr);
+int MPL_gpu_free_host(void *ptr);
+
+int MPL_gpu_host_register(void *ptr, size_t size, unsigned int flags);
+int MPL_gpu_host_unregister(void *ptr);
+int MPL_gpu_host_get_flags(unsigned int *flags, void *ptr);
+
+int MPL_gpu_stream_create(MPL_gpu_stream_t * stream);
```

Comment:
For CUDA current device would be already set, so this interface makes sense. However, for ZE, we need to pass the device when creating a stream (command queue). 
Wondering if we add device as a parameter to this function. Alternatively find a way for ZE code to maintain the current device and access it inside the implementation of MPL_gpu_stream_create. For the latter option, an abstraction such as MPL_gpu_{set,get}_current_device can have benefit. Could such abstraction be helpful for CUDA and HIP?

Reply:
Setting "current device" is not thread safe.

Can we take a step back here?  Even if MPL exposes the stream to MPICH, how will MPICH use it?

Reply:
I think in general MPICH should not set the device as this is typically done by the application before MPI is initialized. It might be good if MPICH could query the `device_id/context` for the process through `MPL` though. One use case would be asynchronous memcpys (`cudaMemcpyAsync`) between host and device (or between devices), for example during intranode communication. In this case we want to be able to create and manage streams internally so that memcpys do not interfere with work on the default stream, that might be used by the application (this would be the case when using `cudaMemcpy`). Offload of collective reductions to the device is another example in which an internally managed stream will be useful (same motivations as above). Am I missing anything?